private/buf/buflsp/buflsp.go

```diff
@@ -0,0 +1,741 @@
+		wellKnownTypesResolver:  wellKnownTypesResolver,
+		wellKnownTypesCachePath: wellKnownTypesCachePath,
+		moduleCachePath:         moduleCachePath,
+	}
+	go func() {
+		for event := range server.fileWatcher.Events {
+			if event.Op&fsnotify.Write == fsnotify.Write {
+				server.lock.Lock()
+				if entry, ok := server.fileCache[event.Name]; ok {
+					if err := server.refreshImage(ctx, entry.resolver); err != nil {
+						server.logger.Sugar().Errorf("failed to build new image: %s", err)
+					}
+				}
+				server.lock.Unlock()
+			}
+		}
+	}()
+	return server, nil
+}
+
+func (s *server) Initialize(ctx context.Context, params *protocol.InitializeParams) (*protocol.InitializeResult, error) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	// Store client info
+	s.folders = params.WorkspaceFolders
+	s.clientCap = params.Capabilities
+
+	// Always load the descriptor.proto file
+	if _, err := s.resolveImport(
+		ctx,
+		s.wellKnownTypesResolver,
+		wellKnownTypesDescriptorProtoPath,
+	); err != nil {
+		return nil, err
+	}
+
+	// Reply with capabilities
+	initializeResult := &protocol.InitializeResult{
+		Capabilities: protocol.ServerCapabilities{
+			TextDocumentSync: &protocol.TextDocumentSyncOptions{
+				OpenClose: true,
+				Change:    protocol.TextDocumentSyncKindFull,
+			},
+			CompletionProvider: &protocol.CompletionOptions{
+				TriggerCharacters: []string{"."},
+			},
+			DocumentFormattingProvider: true,
+			DefinitionProvider:         true,
+			DocumentSymbolProvider:     true,
+			HoverProvider:              true,
+			SemanticTokensProvider:     &protocol.SemanticTokensOptions{},
+		},
+	}
+	return initializeResult, nil
+}
+
+func (s *server) Shutdown(ctx context.Context) error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	return s.fileWatcher.Close()
+}
+
+func (s *server) DidOpen(ctx context.Context, params *protocol.DidOpenTextDocumentParams) error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	filename := normalpath.Normalize(params.TextDocument.URI.Filename())
+
+	// Check if it is already open.
+	if entry, ok := s.fileCache[filename]; ok {
+		entry.refCount++
+		if _, err := entry.updateText(ctx, s, params.TextDocument.Text); err != nil {
+			return err
+		}
+		return nil
+	}
+	var resolver moduleSetResolver
+	if strings.HasPrefix(filename, s.wellKnownTypesCachePath) {
+		resolver = s.wellKnownTypesResolver
+	} else if strings.HasPrefix(filename, s.moduleCachePath) {
+		resolver = newModuleSetResolver(func() (bufmodule.ModuleSet, error) {
+			// Normally, this case won't occur, because the file will already be in the cache.
+			// This case occurs mainly when the LSP is started and a cache file is already open.
+			// We need to recover the module key from the filename in this case.
+			key, _, err := s.cachePathToModuleKey(filename)
+			if err != nil {
+				return nil, err
+			}
+			return s.controller.GetWorkspace(ctx, key.String())
+		})
+	} else {
+		resolver = newModuleSetResolver(func() (bufmodule.ModuleSet, error) {
```

Comment:
I can't understand what is going on here - you are getting a workspace based on a `ProtoFileRef`, which is something I doubt should be used here - this is a local file search thing and has very specific properties. Can you explain this whole block?